libraries/chain/include/eosio/chain/webassembly/eos-vm-oc.hpp

```diff
@@ -329,471 +183,194 @@ struct wasm_function_type_provider<Ret(Args...)> {
+   if constexpr (vm::detail::has_from_wasm_v<T, TC>) {
+      using args_tuple = vm::detail::from_wasm_type_deducer_t<TC, T>;
+      return get_ct_args_one<TC, args_tuple>(std::make_index_sequence<std::tuple_size_v<args_tuple>>());
+   } else {
+      return std::tuple<decltype(make_native_type(std::declval<TC>().as_result(std::declval<T>())))>();
    }
-};
+}
 
-template<bool is_injected, typename WasmSig, typename MethodSig, typename Cls, typename... Params>
-struct intrinsic_function_invoker<is_injected, WasmSig, void, MethodSig, Cls, Params...> {
-   using impl = intrinsic_invoker_impl<is_injected, void_type, std::tuple<Params...>, std::tuple<>>;
+template<typename Args, std::size_t... Is>
+auto get_ct_args(std::index_sequence<Is...>) {
+   return std::tuple_cat(get_ct_args_i<eos_vm_oc_type_converter, std::tuple_element_t<Is, Args>>()...);
+}
 
-   template<MethodSig Method>
-   static void_type wrapper(Params... params) {
+struct result_resolver {
+   template<typename T>
+   auto operator,(T&& res) {
+      return make_native_type(vm::detail::resolve_result(tc, static_cast<T&&>(res)));
+   }
+   eos_vm_oc_type_converter& tc;
+};
+
+template<auto F, typename Interface, typename Preconditions, bool is_injected, typename... A>
+auto fn(A... a) {
+   try {
+      if constexpr(!is_injected) {
+         constexpr int cb_current_call_depth_remaining_segment_offset = OFFSET_OF_CONTROL_BLOCK_MEMBER(current_call_depth_remaining);
+         constexpr int depth_assertion_intrinsic_offset = OFFSET_OF_FIRST_INTRINSIC - (int)boost::hana::index_if(intrinsic_table, ::boost::hana::equal.to(BOOST_HANA_STRING("eosvmoc_internal.depth_assert"))).value()*8;
+         asm volatile("cmpl   $1,%%gs:%c[callDepthRemainOffset]\n"
+                      "jne    1f\n"
+                      "callq  *%%gs:%c[depthAssertionIntrinsicOffset]\n"
+                      "1:\n"
+                      :
+                      : [callDepthRemainOffset] "i" (cb_current_call_depth_remaining_segment_offset),
+                        [depthAssertionIntrinsicOffset] "i" (depth_assertion_intrinsic_offset)
+                      : "cc");
+      }
+      using native_args = vm::flatten_parameters_t<AUTO_PARAM_WORKAROUND(F)>;
+      eosio::vm::native_value stack[] = { a... };
       constexpr int cb_ctx_ptr_offset = OFFSET_OF_CONTROL_BLOCK_MEMBER(ctx);
       apply_context* ctx;
       asm("mov %%gs:%c[applyContextOffset], %[cPtr]\n"
-         : [cPtr] "=r" (ctx)
-         : [applyContextOffset] "i" (cb_ctx_ptr_offset)
-         );
-      (class_from_wasm<Cls>::value(*ctx).*Method)(params...);
-      return void_type();
+          : [cPtr] "=r" (ctx)
+          : [applyContextOffset] "i" (cb_ctx_ptr_offset)
+          );
+      Interface host(*ctx);
+      eos_vm_oc_type_converter tc{&host, eos_vm_oc_execution_interface{stack + sizeof...(A)}};
+      return result_resolver{tc}, eosio::vm::invoke_with_host<F, Preconditions, native_args>(tc, &host, std::make_index_sequence<sizeof...(A)>());
```

Comment:
with clang 10 there are 70+ warnings on this `result_resolver` about `expression result unused`

Reply:
That's unfortunate.